fix: handle semver.io being down or under maintenance

clementprevot opened the pull request:
Do not corrupt the existing versions cache and display easier to understand errors when there has been an issue downloading the NodeJS version from [semver.io](https://semver.io).

For example, today, the 3rd of October, 2022, semver.io is returning:
![image](https://user-images.githubusercontent.com/6450505/193590155-178644ac-e15f-46b6-8655-37535967f603.png)

This case wasn't handled in Nodengine and as my cache was expired, it stored the HTML code of the maintenance page as NodeJS version.  
Then, computing the max matching version failed and it resulted in N being run with `null` as a NodeJS version.  
I then had a weird N error message when running Nodengine (through automatic switching) and make it hard to understand what was happening.

![image](https://user-images.githubusercontent.com/6450505/193590983-f9253d23-f103-4b56-957d-3855a3d681ff.png)

So now, we first properly handle the case where we have an HTML page returned, then we don't corrupt the cache with such HTML page and use the already cached version instead. Finally, we display proper error message to understand where the problem is.

![image](https://user-images.githubusercontent.com/6450505/193591102-0ea441f8-d191-472d-850d-d5e3c8f447ef.png)
![image](https://user-images.githubusercontent.com/6450505/193591057-14788ace-ee97-4da7-b843-813521e6fa8e.png)

Kikobeats commented:
Hello, and thanks for taking care about the project

Although I can see how good is your intention, still the main issue is the fact semver.io is down, so still the code needs a drop-in replacement for semver.io with better uptime 😆

> Also, use the opportunity to add Prettier, reformat the codebase and update the README file (if you don't mind).

I really appreciate the effort, but that is adding so many unnecessary changes that it's hard to just review what's important 😭

clementprevot commented:
> Although I can see how good is your intention, still the main issue is the fact semver.io is down, so still the code needs a drop-in replacement for semver.io with better uptime

Yup, agreed. This PR was just a "quick" try to improve the handling of (not so great) semver.io (to be honest, it's the first time I encounter this issue in more than 7 years using Nodengine 😄)
Do you have any idea of a better way to achieve that without semver.io?

> I really appreciate the effort, but that is adding so many unnecessary changes that it's hard to just review what's important

Indeed, I'm sorry. Let me redo two separated PRs (in fact I added Prettier config because my IDE was force-formatting the file with unwanted changes, and I then got carried away)

Commit:
fix: handle semver.io being down

Do not corrup the existing versions cache and display easier
to understand errors.

clementprevot commented:
@Kikobeats here you go, a first PR easier to understand.
I will then make another one with the Readme, help and code update with Prettier if you agree with that

edmorley commented:
Hi! Semver.io has been deprecated since 2019 and is in the process of being sunset:
https://github.com/heroku/semver.io/issues/74

You will need to switch to another source of node version metadata.

Kikobeats commented:
Since we are using semver.io essentially for listing the available versions:
https://github.com/Kikobeats/nodengine/blob/58fba67cc3544530cc5a69de74b972154e4c344b/bin/fetch.js#L4

Could the solution be as simple as implement that functionality here?
https://github.com/heroku/semver.io/blob/master/lib/sources/node.js#L16

edmorley commented:
> Since we are using semver.io essentially for listing the available versions:

Since semver.io was created, there is now a machine readable version listing published here:
https://nodejs.org/dist/index.json

Release/EOL dates available here:
https://raw.githubusercontent.com/nodejs/Release/main/schedule.json

And also an NPM package that provides an API around these:
https://www.npmjs.com/package/node-release-lines

clementprevot commented:
@Kikobeats would you like me to take a look at these and try to implement something around it?

Kikobeats commented:
@clementprevot yea feel free to commit in this PR 🙂

Commit:
fix: use nodejs.org to retrieve the versions list

clementprevot commented:
It's done @Kikobeats (and it seems to work properly in my computer)

clementprevot commented:
@Kikobeats any chance to review and merge this soon 🤗?

Kikobeats commented:
sorry for the delay; I hope to find some time today to finally review it 🙏

Kikobeats commented:
Thanks! Merged and released under `nodengine@3.5.0`

Can you check all is working as you expect? 🙂

clementprevot commented:
Everything works perfectly!
Thank you!

Are you interested for the cleanup/formatting PR?

Kikobeats commented:
Not sure what does implies. I usually use [standard](https://ghub.io/standard) for code styling.

`EDIT`: registered as issue https://github.com/Kikobeats/nodengine/issues/83 🙂
